src/dune/merlin.ml

```diff
@@ -70,26 +70,34 @@ let quote_for_merlin s =
     s
 
 module Dot_file = struct
-  let b = Buffer.create 256
-
-  let printf = Printf.bprintf b
-
-  let print = Buffer.add_string b
-
-  let to_string ~obj_dirs ~src_dirs ~flags ~pp ~remaindir =
-    let serialize_path = Path.reach ~from:(Path.source remaindir) in
-    Buffer.clear b;
-    print "EXCLUDE_QUERY_DIR\n";
-    Path.Set.iter obj_dirs ~f:(fun p -> printf "B %s\n" (serialize_path p));
-    Path.Set.iter src_dirs ~f:(fun p -> printf "S %s\n" (serialize_path p));
-    Option.iter pp ~f:(printf "%s\n");
-    ( match flags with
-    | [] -> ()
-    | flags ->
-      print "FLG";
-      List.iter flags ~f:(fun f -> printf " %s" (quote_for_merlin f));
-      print "\n" );
-    Buffer.contents b
+  let to_string ~obj_dirs ~src_dirs ~flags ~pp =
+    let serialize_path = Path.to_absolute_filename in
+    let make_directive tag value = Sexp.List [ Atom tag; Atom value ] in
+    let make_directive_of_path tag path =
+      make_directive tag (serialize_path path)
+    in
+    let exclude_query_dir = [ Sexp.List [ Atom "EXCLUDE_QUERY_DIR" ] ] in
+    let obj_dirs =
+      Path.Set.to_list obj_dirs |> List.map ~f:(make_directive_of_path "B")
+    in
+    let src_dirs =
+      Path.Set.to_list src_dirs |> List.map ~f:(make_directive_of_path "S")
+    in
+    let flags =
+      let flags =
+        match flags with
+        | [] -> []
+        | flags ->
+          [ make_directive "FLG"
+              (String.concat ~sep:" " (List.map ~f:quote_for_merlin flags))
```

Comment:
This quoting is a little sad :( I was hopping that we could have gotten rid of it by switching to a structured format. Any chance this could be changed?

FTR, it has been a source of problems in the past, especially on Windows.

Reply:
@trefis believe that it is still (and will be) needed. @trefis can you confirm ?

(If this is something we want to eventually get rid off it would be better to do it now, as it probably require to be a synchronized move between merlin and the readers.)

Reply:
To be precise, what I had in mind was to generate:

```
(FLG (-flag1 -flag2))
```

rather than:

```
(FLG "-flag1 -flag2")
```

Reply:
> If this is something we want to eventually get rid off it would be better to do it now, as it probably requires to be a synchronized move between merlin and the readers

I agree that now is as good a time as any to implement such a change, which I agree we do want indeed.
I haven't looked at this issue (quoting in .merlin) in a long time, so I won't pretend to give advice on what the right format should be, I trust you guys to come up with the right one.
If I remember correctly, the really problematic examples are all related to ppxes which themselves take arguments, but @jeremiedimino will know all about that :grinning: